### PR TITLE
CLN: Simplify map_infer_mask

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This is needed for upcoming numpy string dtype support (getting rid of the ``ndarray[object]`` annotation).

We might also get a decent speedup from using numpy C API functions (but probably not).

EDIT: Looks like benchmarks aren't significantly changed.

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2875,53 +2877,39 @@ def map_infer_mask(
...
+        ndarray result = np.empty(n, dtype=dtype)
+
+        flatiter arr_it = PyArray_IterNew(arr)
+        flatiter result_it = PyArray_IterNew(result)
+
     for i in range(n):
         if mask[i]:
             if na_value is no_default:
-                val = arr[i]
+                val = PyArray_GETITEM(arr, PyArray_ITER_DATA(arr_it))
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:2893]: What is Cython generating in this case? I'm surprised going in to the C API would make a difference

**Review by WillAyd — COMMENTED** [review]
looks good generally

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:2893]: I think this should at least avoid a speed regression taking away the ``ndarray[object]`` annotation.

I assume Cython just does manual Python-style iteration over the elements without the ndarray annotation.

Comment (WillAyd) [reply pandas/_libs/lib.pyx:2893]: I think with the ndarray annotation it would still be using the NumPy iterators. I know you mentioned you aren't seeing a performance difference - I would just revert this unless there is something we see different in the generated code or in the benchmarks

Comment (lithomas1) [reply pandas/_libs/lib.pyx:2893]: The no perf regression is for the new code with numpy C API iteration compared to the ``ndarray[object]`` iteration.

The benchmarks are noisy on my machine, but I'm seeing roughly a 20% slowdown just using regular indexing compared to numpy C API iteration.

File: pandas/core/arrays/string_arrow.py [hunk 2]
```diff
@@ -627,28 +627,25 @@ def _str_map(
                 na_value = np.nan
             else:
                 na_value = False
-            try:
```
Comment (lithomas1) [inline pandas/core/arrays/string_arrow.py:630]: @phofl 

If you're free, could you review the changes to the Arrow strings here?

I think what I have here is correct, but not too familiar with the arrow strings.

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:2893]: Is there an upstream issue in Cython for this?

Comment (lithomas1) [reply pandas/_libs/lib.pyx:2893]: https://github.com/cython/cython/issues/5719 is closely related, and this approach is mentioned there.

I don't know if Cython can do much about this, though. I don't think it's easy to identify when this optimization can be made (esp. for multi-dimensional arrays), but I'm not too much of a compilers person. 

Cython is also encouraging people to use memoryviews anyways.

File: pandas/_libs/lib.pyx [hunk 3]
```diff
@@ -2845,15 +2845,17 @@ no_default = _NoDefault.no_default  # Sentinel indicating the default value.
...
-        object na_value=no_default,
-        cnp.dtype dtype=np.dtype(object)
-) -> "ArrayLike":
+    ndarray arr,
+    object f,
+    const uint8_t[:] mask,
+    *,
+    bint convert=True,
+    object na_value=no_default,
+    object dtype=np.dtype(object)
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:2857]: why this change?  is the new numpy string dtype not an instance of cnp.dtype?

File: pandas/_libs/lib.pyx [hunk 4]
```diff
@@ -2845,15 +2845,17 @@ no_default = _NoDefault.no_default  # Sentinel indicating the default value.
...
+) -> None:
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:2858]: returning None looks inaccurate.  am i reading this wrong?

Comment (lithomas1) [reply pandas/_libs/lib.pyx]: sorry i mixed up the annotations from the old ``_map_infer_mask``

File: pandas/_libs/lib.pyx (hunk 3 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx]: I think this should be an accidental change I forgot to roll back.

In general, is there a reason we prefer ``cnp.dtype``?

With ``cnp.dtype`` I find it harder to do stuff like kind checks on the dtype object.

File: pandas/core/arrays/string_arrow.py (hunk 2 above)
Comment (lithomas1) [reply pandas/core/arrays/string_arrow.py:630]: @mroeschke 

Are you able to review this bit instead?

I'd like to get this in soon.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lithomas1
